### PR TITLE
Allow less strict host header parsing to handle non-RFC2396 compliant host headers

### PR DESCRIPTION
**Pull request by gavinbunney**
The URI parsing changes done in #747 better handled IPv6 addresses, however it introduces issues when host headers contain non-RFC2396 compliant hostnames (for example any `_` or `^` or url-encoded characters).

This PR adds a less strict validation fallback to split-style parsing of the host header when the URI default parsing fails.

### CODE REVIEW

File: zuul-core/src/main/java/com/netflix/zuul/message/http/HttpRequestMessageImpl.java [hunk 1]
```diff
@@ -563,6 +557,48 @@ public Optional<Integer> getClientDestinationPort() {
...
+        }
+
+        try {
+            // attempt to use default URI parsing - this can fail when not strictly following RFC2396,
+            // for example, having underscores in host names will fail parsing
+            URI uri = new URI(/* scheme= */ null, host, /* path= */ null, /* query= */ null, /* fragment= */ null);
+            if (uri.getHost() != null) {
+                return new Pair<>(uri.getHost(), uri.getPort());
+            }
+        } catch (URISyntaxException ignored) { }
```
Comment (artgon) [inline zuul-core/src/main/java/com/netflix/zuul/message/http/HttpRequestMessageImpl.java:580]: Debug logging?

File: zuul-core/src/main/java/com/netflix/zuul/message/http/HttpRequestMessageImpl.java [hunk 2]
```diff
@@ -563,6 +557,48 @@ public Optional<Integer> getClientDestinationPort() {
...
+
+        // fallback to using a colon split
+        // valid IPv6 addresses would have been handled already so any colon is safely assumed a port separator
```
Comment (artgon) [inline zuul-core/src/main/java/com/netflix/zuul/message/http/HttpRequestMessageImpl.java:583]: What if it's an invalid IPv6 address?

File: zuul-core/src/main/java/com/netflix/zuul/message/http/HttpRequestMessageImpl.java [hunk 3]
```diff
@@ -563,6 +557,48 @@ public Optional<Integer> getClientDestinationPort() {
...
+            return new Pair<>(null, -1);
+        }
+
+        String parsedHost = components[0];
+        int parsedPort = -1;
+        if (components.length > 1) {
+            try {
+                parsedPort = Integer.parseInt(components[1]);
+            } catch (NumberFormatException ignored) {
+                // ignore failing to parse port numbers and fallback to default port
```
Comment (artgon) [inline zuul-core/src/main/java/com/netflix/zuul/message/http/HttpRequestMessageImpl.java:596]: Debug logging?

**Comment by artgon** [comment]
I'm not convinced that we want this to be our default functionality. If we do want to allow it, there can be a feature flag that enables it.

File: zuul-core/src/main/java/com/netflix/zuul/message/http/HttpRequestMessageImpl.java (hunk 1 above)
Comment (gavinbunney) [reply zuul-core/src/main/java/com/netflix/zuul/message/http/HttpRequestMessageImpl.java]: Added 🪨

File: zuul-core/src/main/java/com/netflix/zuul/message/http/HttpRequestMessageImpl.java (hunk 2 above)
Comment (gavinbunney) [reply zuul-core/src/main/java/com/netflix/zuul/message/http/HttpRequestMessageImpl.java:599]: There is a fallback below that detects more than 1 colon (e.g. a non-bracketed IPv6 address) which will throw an exception

File: zuul-core/src/main/java/com/netflix/zuul/message/http/HttpRequestMessageImpl.java (hunk 3 above)
Comment (gavinbunney) [reply zuul-core/src/main/java/com/netflix/zuul/message/http/HttpRequestMessageImpl.java:612]: Added 🪨

**Comment by gavinbunney** [comment]
> I'm not convinced that we want this to be our default functionality. If we do want to allow it, there can be a feature flag that enables it.

@artgon Gated this behind a property `zuul.HttpRequestMessage.host.header.strict.validation` which defaults to `true` so it can be selected enabled (or well, disabled)

File: zuul-core/src/test/java/com/netflix/zuul/message/http/HttpRequestMessageImplTest.java [hunk 4]
```diff
@@ -45,6 +49,12 @@
 public class HttpRequestMessageImplTest {
 
     HttpRequestMessageImpl request;
+    private final AbstractConfiguration config = ConfigurationManager.getConfigInstance();
+
+    @After
+    public void resetConfig() {
+        config.clearProperty("zuul.HttpRequestMessage.host.header.strict.validation");
```
Comment (artgon) [inline zuul-core/src/test/java/com/netflix/zuul/message/http/HttpRequestMessageImplTest.java:56]: nice trick!

**Review by artgon — APPROVED** [review]

**Pull request merged**